### DISCUSSION TITLE
Document Approval Workflow workflow ends with HTTP Response activity, but there was no HTTP Request.

### DISCUSSION THREAD

**Discussion by marvel16** [elsa-workflows/elsa-guides#15] [Q&A]
Hello!

Please help to understand how it internally works with HTTP request/response activities:

In this workflow example the workflow is triggered by HTTP request:
https://elsa-workflows.github.io/elsa-core/docs/next/guides/guides-document-approval

![image](https://user-images.githubusercontent.com/5057793/124734054-6ff05880-df1d-11eb-8ab5-55671a186f0f.png)

As i understand when user posted request he wont receive response until steps SetVariable and SendEmail finish gracefully.
1. What will happen with http request if SendEmail fails, because lets say smtp server timed out for 10 min?
Will this activity step will be restarted or workflow will be failed? What will happen with http response which i guess user still waits until http request times out?

2. Don't understand the end of workflow which finishes with HTTP response activity. But where was http request?
Please help understand this =)

Thanks!

**Comment by marvel16** [comment]
Actually I figured out that when you press Approve/Reject that creates an implicit http request triggering the correct branch with activity so that is httpresponse to that.!
But probably documentation has to be clear about which activities have implicit trigger http requests

I'll leave it opened just for some clarifications of how endpoints are dynamically added, my assumption is a middleware that handles any endpoint that is injected via wofklow activity.

**Comment by sfmskywalker** [comment]
> 1. What will happen with http request if SendEmail fails, because lets say smtp server timed out for 10 min?
Will this activity step will be restarted or workflow will be failed? What will happen with http response which i guess user still waits until http request times out?

When executing activities, any exceptions thrown by any activity will be caught by the workflow runner, which will then put the workflow in the `Faulted` state and execution will stop.

When workflows are invoked from the HTTP Endpoint middleware (which triggers workflows starting with **HTTP Endpoint** activities), that middleware will write back a HTTP 500 status code when the workflow faults.

> 2. Don't understand the end of workflow which finishes with HTTP response activity. But where was http request?
Please help understand this =)

When a workflow starts with the **HTTP Endpoint** activity, the way to invoke that workflow is by making an actual HTTP request from an HTTP client. This could be the web browser, Postman, a .NET HTTP Client, etc.

This means that while the workflow executes, it executes within an HTTP context. While being in this context, the **HTTP Response** activity uses it when it's its turn to execute to write back some response.

> Actually I figured out that when you press Approve/Reject that creates an implicit http request triggering the correct branch with activity so that is httpresponse to that.!

Exactly. When a workflow is blocked on the **Signal Received** activity, and that activity is triggered by a web browser invoking the `/signals/trigger/{token}` endpoint (an API controller), then the workflow resumes execution, again within an HTTP context.

> But probably documentation has to be clear about which activities have implicit trigger http requests

Yes. This is an ongoing work in progress. Each activity will be documented in detail.

> my assumption is a middleware that handles any endpoint that is injected via wofklow activity.

That's exactly right. When you do `app.UseHttpActivities` in `Startup`, that's what inserts middleware that tries to match incoming requests against workflows starting with or that are blocked on **HTTP Endpoint** activities.
